### ISSUE TITLE
nn.DataParallel fills None grads with 0

### ISSUE THREAD

**Issue by karandwivedi42** [pytorch/pytorch#7365]
On pytorch 0.4.0, using nn.DataParallel leads to grads of parameters not used being filled with 0. This is a problem when the model has many parameters but only small part of it is used in forward.

Test:
```python
import torch
from torch import nn
class Test(nn.Module):
    def __init__(self):
        super(Test, self).__init__()
        self.l1 = nn.Linear(10,10)
        self.l2 = nn.Linear(10,10)

    def forward(self, input):
        return self.l1(input)

print('CPU:')
net = Test()
inputs = torch.randn(1,10)
outputs = net(inputs)
loss = torch.norm(outputs)
loss.backward()
for j,p in net.named_parameters():
    if p.grad is not None:
        print(j, p.shape, torch.norm(p.grad))
    else:
        print(j)
print('GPU:')
net = Test().to('cuda')
inputs = torch.randn(1,10).to('cuda')
outputs = net(inputs)
loss = torch.norm(outputs)
loss.backward()
for j,p in net.named_parameters():
    if p.grad is not None:
        print(j, p.shape, torch.norm(p.grad))
    else:
        print(j)
print('MultiGPU:')
net = nn.DataParallel(Test().to('cuda'))
inputs = torch.randn(1,10).to('cuda')
outputs = net(inputs)
loss = torch.norm(outputs)
loss.backward()
for j,p in net.module.named_parameters():
    if p.grad is not None:
        print(j, p.shape, torch.norm(p.grad))
    else:
        print(j)
```
For multi GPU we see all grads are filled with zeros. For CPU and GPU the grads are None.

**Comment by karandwivedi42** [comment]
Output:
```
CPU:
l1.weight torch.Size([10, 10]) tensor(3.4795)
l1.bias torch.Size([10]) tensor(1.)
l2.weight
l2.bias
GPU:
l1.weight torch.Size([10, 10]) tensor(2.5339, device='cuda:0')
l1.bias torch.Size([10]) tensor(1., device='cuda:0')
l2.weight
l2.bias
MultiGPU:
l1.weight torch.Size([10, 10]) tensor(2.7430, device='cuda:0')
l1.bias torch.Size([10]) tensor(1.0000, device='cuda:0')
l2.weight torch.Size([10, 10]) tensor(0., device='cuda:0')
l2.bias torch.Size([10]) tensor(0., device='cuda:0')
```

**Comment by apaszke** [comment]
Why is this a problem? We could surely fix that, but I'm afraid that it would make DataParallel slower, because it would need to actually look into what the gradients are.

**Comment by fmassa** [comment]
We should also consider that this makes the optimized slower I believe, and in some cases `None` in the gradient means that the update is entirely skipped, and no momentum / weight decay is applied, so there is some difference

**Comment by karandwivedi42** [comment]
I came across this implementing something like [PathNet](https://arxiv.org/abs/1701.08734) where there is a very large network and we choose only a small part of it at each forward pass.

**Comment by karandwivedi42** [comment]
@apaszke @fmassa I would be glad to take this up if you can give some pointers. I am getting around 2x slowdown when using 4 GPUs compared to single GPU.

**Comment by apaszke** [comment]
I don't think we can resolve this without knowing the set of parameters you'll use in `forward` upfront. The problem is in `broadcast_coalesced`, which will compact them into blocks quite arbitrarily, and as long as a single parameter gets grad, all other will get zeros.

**Comment by ZhengRui** [comment]
I am having the same issue, my use case is when only fine-tuning a small head attached to a much bigger backbone network 

```python
with torch.no_grad():
    o = backbone(x)

o = head(o)
```

in single gpu case, after backward, backbone parameter's gradients are None, while in multi gpu cases, their gradients are all 0s, I am not sure if it does backward operations on the backbone or not, even if not  it wastes some gpu memory, so I would like it to behave as in single gpu case.
